pmlc/target/intel_gen/affine_pack.cc

```diff
@@ -0,0 +1,144 @@
+      auto lbExpr =
+          op.lowerBoundsMap().getResult(i).dyn_cast<AffineConstantExpr>();
+      if (!lbExpr || lbExpr.getValue() != 0 || op.getSteps()[i] != 1 ||
+          ranges[i] == 1) {
+        op.emitError("Unable to reduce dims to 3 for non-normalized loop");
+        return failure();
+      }
+    }
+    // Initialize packing state
+    SmallVector<int64_t, 3> curPack = {1, 1, 1};
+    SmallVector<IndexPacking, 6> packInfo(ranges.size());
+    // Pack all po2 indexes into index #0
+    for (unsigned i = 0; i < ranges.size(); i++) {
+      if (IsPo2(ranges[i])) {
+        packInfo[i].sourceIdx = 0;
+        packInfo[i].floorDiv = curPack[0];
+        packInfo[i].mod = ranges[i];
+        curPack[0] *= ranges[i];
+      }
+    }
+    // Now pack all non-po2 into seperate indexes if possible (and spill the
+    // rest to #2)
+    unsigned curIdx = 0;
+    for (unsigned i = 0; i < ranges.size(); i++) {
+      if (!IsPo2(ranges[i])) {
+        packInfo[i].sourceIdx = curIdx;
+        packInfo[i].floorDiv = curPack[curIdx];
+        packInfo[i].mod = ranges[i];
+        curPack[curIdx] *= ranges[i];
+        curIdx = std::min(curIdx + 1, 2u);
+      }
+    }
+    // Remove extra indexes
+    while (curPack.size() > 0 && curPack.back() == 1) {
+      curPack.pop_back();
+    }
+    // Make a noralized affineParallel
+    auto builder = OpBuilder(op);
+    auto newLoop = builder.create<AffineParallelOp>(
+        op.getLoc(),
+        /*resultsTypes=*/ArrayRef<Type>{},
+        /*reductions=*/ArrayRef<AtomicRMWKind>{},
+        /*ranges=*/curPack);
+    // Make an affine apply for each original index
+    auto ibuild = newLoop.getBodyBuilder();
+    for (unsigned i = 0; i < ranges.size(); i++) {
+      const auto &info = packInfo[i];
+      // Begin with the original value
+      AffineExpr cur = getAffineDimExpr(0, op.getContext());
+      // Divide by floorDiv (== 1 case should be canonicalized away)
+      cur = cur.floorDiv(info.floorDiv);
+      // If needed, do a modulus
+      if (info.floorDiv * info.mod !=
+          static_cast<uint64_t>(curPack[info.sourceIdx])) {
+        cur = cur % info.mod;
+      }
+      // Make an affine map
+      auto map = AffineMap::get(1, 0, cur);
+      // Make the affine apply
+      Value newIV = newLoop.getIVs()[info.sourceIdx];
+      Value mapped =
+          ibuild.create<AffineApplyOp>(op.getLoc(), map, ValueRange{newIV});
+      op.getIVs()[i].replaceAllUsesWith(mapped);
+    }
+    // Slice in the interior
+    newLoop.getBody()->getOperations().splice( //
+        std::prev(newLoop.getBody()->end()),   //
+        op.getBody()->getOperations(),         //
+        op.getBody()->begin(),                 //
+        std::prev(op.getBody()->end()));
+    newLoop.setAttr("hardware", hardware);
+    // Erase old op
+    op.erase();
+    return success();
+  }
+
+  void runOnFunction() override {
+    getFunction().walk([&](AffineParallelOp op) {
```

Comment:
Maybe you want `getOps<AffineParallelOp>()` so that this isn't recursive?

Reply:
It's intended to be recursive (Thread loops can totally be inside Block loops)